### DISCUSSION TITLE
Related to configurations file backup

### DISCUSSION THREAD

**Discussion by zk637** [tnodir/fort#252] [Q&A]
I want to back up a copy of the configuration file via hard link, but there are more files than "Options: Backup" in the path to "My Fort Friewall: Statistics"

I don't know if implementing "Options: Backup" requires hard links to those files?

Additional information, "My Fort Friewall: Statistics" The path to Statistics does not contain "FortFirewall.user.ini"

![image](https://github.com/tnodir/fort/assets/78423939/00c66531-f519-4ebd-939d-b37b1d73c36a)

"Options: Backup Exported file" :
![image](https://github.com/tnodir/fort/assets/78423939/3feaa8ee-a88b-4d71-987f-3f9bc16d3922)

**Comment by tnodir** [comment]
Options: "Backup" copies only settings files.

Do you want to also copy statistics files?

**Reply by zk637** [reply]
Yep, I want to back up all files except the temporary data generated at runtime. It's like a snapshot

**Reply by tnodir** [reply]
Ok, created #253

**Reply by zk637** [reply]
Ok,Thanks. Do you plan to develop it as a feature? I mean version 3.12.10 statistics files can't be backed up in real time by hard link alone?

**Reply by tnodir** [reply]
> Do you plan to develop it as a feature?

Yes, later.

> I mean version 3.12.10 statistics files can't be backed up in real time by hard link alone?

You can backup by using `sqlite3.exe`:
- download `sqlite-tools-win-x64-XXX.zip` from https://sqlite.org/download.html
- run `"[PATH-TO-SQLITE]\sqlite3.exe" "[PATH-TO-PROFILE]\FortFirewall.stat" ".backup [PATH-TO-BACKUP]\FortFirewall.stat"`

**Reply by zk637** [reply]
Well, I want it to be optional when exporting ,This is Netlimiter's statistic, and it's Frighteningly big

![image](https://github.com/tnodir/fort/assets/78423939/878fd846-e251-4baf-a72c-dfda551653e2)

**Reply by tnodir** [reply]
> I want it to be optional when exporting

Then it's better to export/import from Statistics window. Renamed the #253.

**Reply by zk637** [reply]
There is a question, if the exported file is a fixed time node, how to ensure that it is changed in real time? Because my understanding is to separate the statistics into a file and then back it up with a hard link

**Comment by tnodir** [comment]
> "My Fort Friewall: Statistics" The path to Statistics does not contain "FortFirewall.user.ini"

"FortFirewall.user.ini" is from "C:\Users\\[USER-NAME]\AppData\Local\Fort Firewall\".

**Comment by zk637** [comment]
> I want to back up a copy of the configuration file via hard link, but there are more files than "Options: Backup" in the path to "My Fort Friewall: Statistics"
> 
> I don't know if implementing "Options: Backup" requires hard links to those files?
> 
> ![image](https://private-user-images.githubusercontent.com/78423939/332442967-00c66531-f519-4ebd-939d-b37b1d73c36a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyOTg2OTQsIm5iZiI6MTcxNjI5ODM5NCwicGF0aCI6Ii83ODQyMzkzOS8zMzI0NDI5NjctMDBjNjY1MzEtZjUxOS00ZWJkLTkzOWQtYjM3YjFkNzNjMzZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIxVDEzMzMxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyNDYxN2QxMDdmNTQ3NTkxZGVmY2EyMTNjZjllMWQyM2RlNGM4ZGFkMTE0YTdiMWJjMWViMWRmMmJiMzRjNDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.HEKypceFd2bkgABzlJ8m06UPYjN95yjyaCBcuZZv5Kc)
> 
> "Options: Backup Exported file" : ![image](https://private-user-images.githubusercontent.com/78423939/332444182-3feaa8ee-a88b-4d71-987f-3f9bc16d3922.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyOTg2OTQsIm5iZiI6MTcxNjI5ODM5NCwicGF0aCI6Ii83ODQyMzkzOS8zMzI0NDQxODItM2ZlYWE4ZWUtYTg4Yi00ZDcxLTk4N2YtM2Y5YmMxNmQzOTIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIxVDEzMzMxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE5ZTk0YzU2MzNkYjRlYTY0MzViYWQzY2Y5OTdhZWY3NzZiNzFlYjYzZjA1ZWQ3OTFhMTkwYTljZDMyNDQ2ZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.N_uOf7nHxrtNbIeb08TD-s6EqMJDSU6IiRs4gl6sr14)



> Options: "Backup" copies only settings files.选项：“备份”仅复制设置文件。
> 
> Do you want to also copy statistics files?您还想复制统计文件吗？



> Ok, created #253 好的，已创建 #253



> > Do you plan to develop it as a feature?您打算将其开发为一项功能吗？
> 
> Yes, later. 是的，稍后。
> 
> > I mean version 3.12.10 statistics files can't be backed up in real time by hard link alone?我的意思是3.12.10版本统计文件不能单独通过硬链接实时备份？
> 
> You can backup by using `sqlite3.exe`:您可以使用 `sqlite3.exe` 进行备份：
> 
> * download `sqlite-tools-win-x64-XXX.zip` from https://sqlite.org/download.html从 https://sqlite.org/download.html 下载 `sqlite-tools-win-x64-XXX.zip`
> * run `"[PATH-TO-SQLITE]\sqlite3.exe" "[PATH-TO-PROFILE]\FortFirewall.stat" ".backup [PATH-TO-BACKUP]\FortFirewall.stat"` 运行 `"[PATH-TO-SQLITE]\sqlite3.exe" "[PATH-TO-PROFILE]\FortFirewall.stat" ".backup [PATH-TO-BACKUP]\FortFirewall.stat"`

Well, I want it to be optional when exporting

> > Do you plan to develop it as a feature?您打算将其开发为一项功能吗？
> 
> Yes, later. 是的，稍后。
> 
> > I mean version 3.12.10 statistics files can't be backed up in real time by hard link alone?我的意思是3.12.10版本统计文件不能单独通过硬链接实时备份？
> 
> You can backup by using `sqlite3.exe`:您可以使用 `sqlite3.exe` 进行备份：
> 
> * download `sqlite-tools-win-x64-XXX.zip` from https://sqlite.org/download.html从 https://sqlite.org/download.html 下载 `sqlite-tools-win-x64-XXX.zip`
> * run `"[PATH-TO-SQLITE]\sqlite3.exe" "[PATH-TO-PROFILE]\FortFirewall.stat" ".backup [PATH-TO-BACKUP]\FortFirewall.stat"` 运行 `"[PATH-TO-SQLITE]\sqlite3.exe" "[PATH-TO-PROFILE]\FortFirewall.stat" ".backup [PATH-TO-BACKUP]\FortFirewall.stat"`



> > I want it to be optional when exporting我希望导出时它是可选的
> 
> Then it's better to export/import from Statistics window. Renamed the #253.那么最好从统计窗口导出/导入。重命名了 #253 。



> > "My Fort Friewall: Statistics" The path to Statistics does not contain "FortFirewall.user.ini"“My Fort Friewall：统计”统计路径不包含“FortFirewall.user.ini”
> 
> "FortFirewall.user.ini" is from "C:\Users\[USER-NAME]\AppData\Local\Fort Firewall".“FortFirewall.user.ini”来自“C:\Users\[用户名]\AppData\Local\Fort Firewall”。

There is a question, if the exported file is a fixed time node, how to ensure that it is changed in real time? Because my understanding is to separate the statistics into a file and then back it up with a hard link

**Reply by tnodir** [reply]
> if the exported file is a fixed time node, how to ensure that it is changed in real time?

I don't understand the question.

You can search (e.g. by Google) for the following queries:
- "how to backup sqlite database"
- "how to replicate sqlite database"

Also see https://litestream.io

**Reply by zk637** [reply]
No,no, I mean is it possible to separate statistical analysis from other files into one or more files? All you need to do is create a hard connection to these files.

**Reply by tnodir** [reply]
> I mean is it possible to separate statistical analysis from other files into one or more files?

The `FortFirewall.stat` file contains the statistics. You can not just copy the file, because it is a SQLite database.

`FortFirewall.stat-shm` and `FortFirewall.stat-wal` are temporary files.

**Reply by zk637** [reply]
I checked https://litestream.io/reference/replicate/ Mentioned configuration file ,If copy that as well ,Whether backup can be implemented,Because I understand that FortFirewall.stat and nlstatsV4.db are similar. And sorry for my inaccurate description of "snapshot". 
Finally, I want to make sure that I can copy the above files to achieve the backup of statistical data?

**Reply by tnodir** [reply]
> I want to make sure that I can copy the above files to achieve the backup of statistical data?

Just use `sqlite3.exe`: https://github.com/tnodir/fort/discussions/252#discussioncomment-9509088

**Reply by zk637** [reply]
The reason I don't want to use this method is because it introduces a new dependency on sqlite3.exe, and links use reference counting is a good implementation if you just need to create copies of files without requiring a third party. Hard link reference:https://en.wikipedia.org/wiki/Hard_link

**Reply by tnodir** [reply]
> links use reference counting is a good implementation if you just need to create copies of files without requiring a third party.

I was wondering which "hard link" do you mean.

But it's still dangerous. Because you can not just copy that file to other machine - SQLite manages the file with other temporary files (-shm, -wal).

**Reply by zk637** [reply]
Maybe that's it, because I just want to back up my statistics,You seem to forget that hard connections cannot cross partitions,I have now used symbolic links

![image](https://github.com/tnodir/fort/assets/78423939/38a2ce50-6ee3-4fd1-b26c-3e3430c0c318)


![image](https://github.com/tnodir/fort/assets/78423939/15f6fc63-340b-4732-baf3-45846daf1605)

**Reply by zk637** [reply]
I think the problem is over, there is no need to stop Fort Firewall when creating a hard link, if only FortFirewall.stat and fortfirewall.stat-block backup statistics are needed.However, the use of symbolic links still needs to stop the Fort Firewall, I think this part can be improved?
![image](https://github.com/tnodir/fort/assets/78423939/b23ffdf8-fcdb-4491-9f37-aa84a805a0c4)

![image](https://github.com/tnodir/fort/assets/78423939/06fc005f-f028-4168-b48c-684b7f52574f)

**Reply by tnodir** [reply]
Again, you can not copy that files in parallel of working Fort Firewall. Directly or via links, it does not matter.

Please wait for #253.

**Reply by zk637** [reply]
I want to share the latest progress, I have written the vbs script to implement the hard connection as a normal file so that it can be robocopy for cross-file system backup! But I won't be able to compare the hash to see if the two matches until you've safely released the resource. And I want to know if there is a security risk? I used a hard link to access FortFirewall.stat

![image](https://github.com/tnodir/fort/assets/78423939/3c520e15-0aa4-4def-abee-6d9d74e2f577)

![image](https://github.com/tnodir/fort/assets/78423939/7e0e7bf6-d43e-451e-91b1-db1770e4aab0)

![image](https://github.com/tnodir/fort/assets/78423939/581f47e2-f7f3-44cb-b328-3dd7129cd29c)

![image](https://github.com/tnodir/fort/assets/78423939/15101572-7014-427c-b9be-eb8346f1a6b8)

**Comment by zk637** [comment]
Update since 3.12.11 will automatically stop the service, so that I can determine the hash value of the data at the key node, so I think the problem should be solved?(Because I'm not sure the logic of the program needs to be asked)

![image](https://github.com/tnodir/fort/assets/78423939/b15c9daf-bb01-4120-87ed-deb6cc026bbc)

**Comment by zk637** [comment]
The problem has been preliminarily resolved, see details

> https://github.com/tnodir/fort/issues/257#issuecomment-2134908985
